### ISSUE TITLE
Login over OpenID can't download objects

### ISSUE THREAD

**Issue by mieseprem** [minio/minio#12381]
Our MinIO Service is connected to Keycloak by setting 
- MINIO_IDENTITY_OPENID_CONFIG_URL
- MINIO_IDENTITY_OPENID_CLIENT_ID
- MINIO_IDENTITY_OPENID_CLAIM_NAME

We have not configured any users in MinIO except the local admin account that is "created" by setting
- MINIO_ACCESS_KEY
- MINIO_SECRET_KEY

As of RELEASE.2021-05-11T23-27-41Z the users authorized over OpenID are no longe able to download objects.

## Expected Behavior
A User click on "OpenID", gets redirected to Keycloak Login page, gets back to MinIO, click on desired bucket, click on object, click on "Download object", download object
Note: up to RELEASE.2021-04-22T15-44-28Z the "Expected Behavior" worked as expected.

## Current Behavior
Related to all realease from (including) RELEASE.2021-05-11T23-27-41Z to (including) RELEASE.2021-05-27T22-06-31Z
A User click on "OpenID", gets redirected to Keycloak Login page, gets back to MinIO, click on desired bucket, click on object, click on "Download object" -> "The access key ID you provided does not exist in our records"

![grafik](https://user-images.githubusercontent.com/50367822/119786917-f5b9c680-bed0-11eb-920f-b1017be3bc13.png)

## Possible Solution
Maybe something mentioned in https://github.com/minio/minio/releases/tag/RELEASE.2021-05-11T23-27-41Z is a breaking change but I don't get what (if it's not a bug)

## Steps to Reproduce (for bugs)

1. created a working instance
2.1. docker run --rm -d --name minio-0422 -p 9001:9000 -e MINIO_ACCESS_KEY=acckey -e MINIO_SECRET_KEY=12345678 -e M
INIO_IDENTITY_OPENID_CONFIG_URL=https://keycloak.url/.well-known/openid-configuration -e MINIO_IDENTITY_OPENID_CLIENT_ID=myClientId -e
 MINIO_IDENTITY_OPENID_CLAIM_NAME=myPolicy minio/minio:RELEASE.2021-04-22T15-44-28Z server /data 
2. created a non-working instance
2.1. docker run --rm -d --name minio-0526 -p 9002:9000 -e MINIO_ACCESS_KEY=acckey -e MINIO_SECRET_KEY=12345678 -e M
INIO_IDENTITY_OPENID_CONFIG_URL=https://keycloak.url/.well-known/openid-configuration -e MINIO_IDENTITY_OPENID_CLIENT_ID=myClientId -e
 MINIO_IDENTITY_OPENID_CLAIM_NAME=myPolicy minio/minio:RELEASE.2021-05-26T00-22-46Z server /data

I traced with 'mc admin trace -v -a <TARGET>' but there is no error reported. All is 200 and OK.

## Context
We have users that should download reportd provided on MinIO service but are unable to if we upgade MinIO to latest version.

## Regression
Sorry, but I don't know.

## Your Environment
We run MinIO as docker image. But added own certificates:

```Dockerfile
FROM minio/minio:RELEASE.2021-05-26T00-22-46Z

COPY files/ca-certs/* /etc/pki/ca-trust/source/anchors/
RUN update-ca-trust
```

**Comment by irudayambics** [comment]
I am getting the same error.
[
<img width="720" alt="access key id" src="https://user-images.githubusercontent.com/3430078/119918444-0d5d8180-bf9b-11eb-927b-a988354f1698.PNG">
](url)

**Comment by harshavardhana** [comment]
Have you tried using https://github.com/minio/console  ?

**Comment by irudayambics** [comment]
I have not tried minio console project and let me try that. Could you clarify the following?

1. Does minio console also need same environment variables like minio object browser? Because this minio console project did not mention anything about keycloak integration.
2. Is minio console a replacement project for minio object browser UI?
3. Do we still need minio object browser UI after configuring minio console project?

**Comment by harshavardhana** [comment]
> I have not tried minio console project and let me try that. Could you clarify the following?
> 
> 1. Does minio console also need same environment variables like minio object browser? Because this minio console project did not mention anything about keycloak integration.
> 2. Is minio console a replacement project for minio object browser UI?
> 3. Do we still need minio object browser UI after configuring minio console project?

No you need to follow console docs for that..

**Comment by irudayambics** [comment]
Thanks @harshavardhana. How do I install binary release on my cent OS?

**Comment by harshavardhana** [comment]
> Thanks @harshavardhana. How do I install binary release on my cent OS?

Look at the releases here https://github.com/minio/console/releases

**Comment by mieseprem** [comment]
> Have you tried using https://github.com/minio/console ?

Not yet, I will give it a try.

**Comment by mieseprem** [comment]
I reproduced the issue while console was connected to minio. I hoped to get more insight into the issue but I don't.
After I provoked the "The access key ID you provided ..." error I run the Diagnostic in Console.

I don't find any usefull information in it, but maybe you will: [diagnostic.zip](https://github.com/minio/minio/files/6584202/diagnostic.zip)

**Comment by irudayambics** [comment]
@harshavardhana I have tried to setup minio console using the link below.

https://github.com/minio/console#connect-console-to-a-minio-using-tls-and-a-self-signed-certificate

My console is running on http://<*******************>:9090/login and I can see console login page. After I input console secret key and access key, below error is shown. Could you assist me? Also please share any website link that has all the environment variables used by minio console project. 

<img width="588" alt="error" src="https://user-images.githubusercontent.com/3430078/120987342-14ad3800-c7b0-11eb-89a4-15068df5a7b0.PNG">

Minio console environment variables are below.

#Salt to encrypt JWT payload
CONSOLE_PBKDF_PASSPHRASE=SECRET

#Required to encrypt JWT payload
CONSOLE_PBKDF_SALT=SECRET

#MinIO Endpoint
CONSOLE_MINIO_SERVER=https://***************:9000

#MinIO User Access Key (used for Console Login)
CONSOLE_ACCESS_KEY=console

#MinIO User Secret Key (used for Console Login)
CONSOLE_SECRET_KEY=Console

MCS_MINIO_SERVER_TLS_ROOT_CAS=/etc/minio/certs/CAs/public1.crt

#Minio to the configuration directory
#CONSOLE_OPTS="--host **.**.**.com --port 9090 --tls-host **.**.**.com --tls-port 9445 --certs-dir /etc/minio-console/certs --tls-certificate /etc/minio-console/certs/public.crt --tls-key /etc/minio-console/certs/private.key"

**Comment by irudayambics** [comment]
I resolved it. Actually --certs-dir, --tls-certificate and --tls-key parameters did not work also looks like the source code hardcoded with .console directory to look for certificates. I created ~/.console/certs directory under user who started my cent OS minio console service and moved SSL certs and CAs. 

I am now able to login to minio console using form login.

However when minio is configured with below IDP environment variables, login page keeps loading with error message "An error occurred, please try again" Any solution on this?

<img width="720" alt="keep loading idp login" src="https://user-images.githubusercontent.com/3430078/121119655-981d6680-c84e-11eb-98ae-300050bd4b64.PNG">

**Env Variables**

CONSOLE_IDP_URL=https://**:8443/auth/realms/**/.well-known/openid-configuration
CONSOLE_IDP_CLIENT_ID=minio-console
CONSOLE_IDP_SECRET=e25ac469-cf5e-4e50-9e52-fbda460c7967
CONSOLE_IDP_CALLBACK=https://*********:9445/oauth_callback

I loaded CA cert of my keycloak under .console/certs/CAs directory.

**Comment by thebrooder** [comment]
I experienced the exact same problem with oauth/keycloak configuration - upgraded minio yesterday to binary RELEASE.2021-06-07T21-40-51Z and all downloads failed after with `The access key ID you provided does not exist in our records`.   Status 200 in the trace.  No issue with the same users being able to upload files and create new buckets, same user can upload a file and then not be able to download the same file.  No errors are logged in either minio or keycloak
I downgraded to binary RELEASE.2021-04-22T15-44-28Z and the problem goes away.   Locally created users, created with `mc admin` are able to download files fine on both releases.

**Comment by irudayambics** [comment]
@thebrooder How did you configure keycloak CA cert to the downloaded minio console project? I have no idea of where to place keycloak CA cert for my my minio console project. I know minio console looks for SSL certs under .console/certs directory. There is no document on adding external CA to minio console projec?

**Comment by harshavardhana** [comment]
> @thebrooder How did you configure keycloak CA cert to the downloaded minio console project? I have no idea of where to place keycloak CA cert for my my minio console project. I know minio console looks for SSL certs under .console/certs directory. There is no document on adding external CA to minio console projec?

You need to put the trust certs in `~/.console/certs/CAs`

**Comment by irudayambics** [comment]
@harshavardhana I have already added trust certs in ~/.console/certs/CAs. I see access minio console UI on port 9443 but where to place keycloaks CA cert because I want to enable SSO with key cloak? 

Below is my configuration.
<img width="312" alt="ca certs" src="https://user-images.githubusercontent.com/3430078/121466506-1cf0b780-c9ea-11eb-97e8-1bf11081f742.PNG">

**Comment by harshavardhana** [comment]
> @harshavardhana I have already added trust certs in ~/.console/certs/CAs. I see access minio console UI on port 9443 but where to place keycloaks CA cert because I want to enable SSO with key cloak?

You should put it in in `~/.console/certs/CAs`  just name it whatever it can be..

**Comment by harshavardhana** [comment]
CAs directory can have as many public certificates as you want just name them as `file.crt, file2.crt ...` and so on..

**Comment by irudayambics** [comment]
@harshavardhana I added many public CAs and keycloak's is one of them. Please see the screenshot below.

<img width="305" alt="multiple certs" src="https://user-images.githubusercontent.com/3430078/121469167-9be7ef00-c9ee-11eb-9921-0cbc34c61a29.PNG">

I can login to minio console 

But I still could not see Login with SSO button after enabling IDP environment variables.  The error is "An error occurred, please try again" and the screenshot is below.

<img width="720" alt="eror" src="https://user-images.githubusercontent.com/3430078/121470156-57f5e980-c9f0-11eb-8d41-bdd9845544cd.PNG">

Backend log is as below. 
<img width="597" alt="back end" src="https://user-images.githubusercontent.com/3430078/121470298-97243a80-c9f0-11eb-8e18-79c433769d41.PNG">

**Comment by harshavardhana** [comment]
You need to also trust MinIO server certs @irudayambics from console

**Comment by mieseprem** [comment]
Can we please keep focus on the issue itself and not discussing setup console?
This is regarding OpenId authenticated users that are not confugured via 'mc admin user' and can do all but download objects.

Is there a way to get closer to the cause without using the console? I think we all have the ability to use 'mc'

**Comment by irudayambics** [comment]
@harshavardhana I have added Minio server's CA certificate and still unable to see Login with SSO button.

Screenshot of cert directory
<img width="318" alt="still error" src="https://user-images.githubusercontent.com/3430078/121486712-f8a1d480-ca03-11eb-99f8-f6b3982ab953.PNG">

Screenshot of backend 
<img width="595" alt="backend still errior" src="https://user-images.githubusercontent.com/3430078/121486753-035c6980-ca04-11eb-9cd9-a26dfe896526.PNG">

**Comment by thebrooder** [comment]
I have setup the console authenticating with a local minio account running  from localhost.  It does not show anything in its logs either about the failed downloads.  I can dump the diagnostics if you like, but that seems to just dump the current running configuration.   I can't test the download issue with console itself as the failed downloads only affect oauth authorized accounts and Console does not appear to support OAuth (or at least its not documented anywhere).

**Comment by mieseprem** [comment]
I can confirm that as of RELEASE.2021-06-14T01-29-23Z downloading of objects by remote users is working again.

**Comment by mieseprem** [comment]
Hi @harshavardhana - Just a follow-up message on this topic:

This issue is not fully fixed with #12489

If a user is authenticated via OpenId all is fine now. But only UNTIL the MinIO Pod is deleted/recreated but the Token is still valid. The user doesn't get something like "Session invalid, you need to re-authenticate" but is still able to browse buckets and list objects in them. But if the user tries to download the known error appears.
